src/blockchain/rpc.rs

```diff
@@ -695,81 +701,54 @@ where
+fn list_transactions(
+    client: &Client,
     page_size: usize,
-    page_index: usize,
-
-    stack: Vec<ListTransactionResult>,
-    done: bool,
-}
-
-impl<'a> CoreTxIter<'a> {
-    fn new(client: &'a Client, mut page_size: usize) -> Self {
-        if page_size > 1000 {
-            page_size = 1000;
-        }
-
-        Self {
-            client,
-            page_size,
-            page_index: 0,
-            stack: Vec::with_capacity(page_size),
-            done: false,
-        }
-    }
-
-    /// We want to filter out conflicting transactions.
-    /// Only accept transactions that are already confirmed, or existing in mempool.
-    fn keep_tx(&self, item: &ListTransactionResult) -> bool {
-        item.info.confirmations > 0 || self.client.get_mempool_entry(&item.info.txid).is_ok()
+) -> Result<impl Iterator<Item = ListTransactionResult>, Error> {
+    if !(1..=1000).contains(&page_size) {
+        return Err(Error::Generic(format!(
+            "Core RPC method `listtransactions` must have `page_size` in range [1 to 1000]: got {}",
+            page_size
+        )));
     }
-}
-
-impl<'a> Iterator for CoreTxIter<'a> {
-    type Item = Result<ListTransactionResult, Error>;
-
-    fn next(&mut self) -> Option<Self::Item> {
-        loop {
-            if self.done {
-                return None;
-            }
-
-            if let Some(item) = self.stack.pop() {
-                if self.keep_tx(&item) {
-                    return Some(Ok(item));
-                }
-            }
 
-            let res = self
-                .client
-                .list_transactions(
-                    None,
-                    Some(self.page_size),
-                    Some(self.page_size * self.page_index),
-                    Some(true),
-                )
-                .map_err(Error::Rpc);
-
-            self.page_index += 1;
-
-            let list = match res {
-                Ok(list) => list,
-                Err(err) => {
-                    self.done = true;
-                    return Some(Err(err));
-                }
-            };
-
-            if list.is_empty() {
-                self.done = true;
-                return None;
+    // helper for `.take_while` so that we can obtain the first error
+    // TODO: remove when we can use `.map_while`
+    let mut got_err = false;
+
+    // obtain results in batches (of `page_size`)
+    let nested_list = (0_usize..)
+        .map(|page_index| {
+            client.list_transactions(
+                None,
+                Some(page_size),
+                Some(page_size * page_index),
+                Some(true),
+            )
+        })
+        // take until returned rpc call is empty or until error
+        // TODO: replace with the following when MSRV is 1.57.0:
+        // `.map_while(|res| res.map(|l| if l.is_empty() { None } else { Some(l) }).transpose())`
+        .take_while(|res| {
+            if got_err || res.is_ok() && res.as_ref().unwrap().is_empty() {
```

Comment:
I'm not sure if that works, but maybe you can replace the second condition with something like `matches!(res, Ok(&[]))`.

But if you prefer it this way, that also works for me. Up to you!